### DISCUSSION TITLE
Some questions about hole punching

### DISCUSSION THREAD

**Discussion by mkuhn19** [libp2p/rust-libp2p#4388] [Q&A]
Hi,

I am trying to develop a peer to peer application using libp2p.
I am currently implementing the hole punching aspect and have several questions :

1. Is there a way to run the [hole punching example](https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html) inside Docker containers on the same machine ?
  
2. Is there an automatic fallback when hole punch fails ? In the [specs](https://docs.rs/libp2p-request-response/latest/libp2p_request_response/struct.Behaviour.html#method.send_request), it say that the relay connection is still used but this is not what I observe in practice : the connection between the 2 peers is closed after several attempts.
  
3. I am using request response to send message from a peer to another. When calling send_request, it tries to dial the remote peer and when there is a connection, it sends a message, as described [here](https://docs.rs/libp2p-request-response/latest/libp2p_request_response/struct.Behaviour.html#method.send_request). The issue is that, when a relayed circuit is established between 2 peers, the request will be sent directly and not wait until hole punching is performed. Can I tell request response to wait before sending the message ?

Thank you !

**Comment by thomaseizinger** [comment]
> Is there a way to run the hole punching example inside Docker containers on the same machine ?

I am not sure what that would achieve. To test hole punching, you need two NATs. Inside a docker container, you only have a single interface by default.

I am working on hole punching tests at the moment but those will likely run on bare Linux because I need to set up network namespaces myself.

> Is there an automatic fallback when hole punch fails ? In the specs, it say that the relay connection is still used but this is not what I observe in practice : the connection between the 2 peers is closed after several attempts.

If you are using the connection for something else, it should not be closed I believe.

> I am using request response to send message from a peer to another. When calling send_request, it tries to dial the remote peer and when there is a connection, it sends a message, as described here. The issue is that, when a relayed circuit is established between 2 peers, the request will be sent directly and not wait until hole punching is performed. Can I tell request response to wait before sending the message ?

You can wait for the hole-punch event and only then send the message although I don't think we currently have an API to specify, which connection the message should be sent on.

You could manually close the relayed connection upon success and only then issue the request. It will then be sent on the only remaining, direct connection.

**Reply by mkuhn19** [reply]
> I am not sure what that would achieve. To test hole punching, you need two NATs. Inside a docker container, you only have a single interface by default. 
 I am working on hole punching tests at the moment but those will likely run on bare Linux because I need to set up network namespaces myself.

Okay that's good to know. How would you recommend testing an application ? I have business logic between my peers and I currently use docker to quickly start a network composed of some amount of peers. I can then check the logs to see if everything went well. However, because I can't achieve hole punching, I don't know how I can continue testing my application.

> If you are using the connection for something else, it should not be closed I believe.

I agree with you however the dialer peer does not get a `DirectConnectionUpgradeFailed` event. Do I need to manually count the `OutgoingConnectionError` to the listener peer to know when I should use the relayed connection ? And how can I use the relayed connection as the hole punching attempt is automatically made whenever a relay circuit is opened to a peer that also supports dcutr ?

> You can wait for the hole-punch event and only then send the message although I don't think we currently have an API to specify, which connection the message should be sent on.
You could manually close the relayed connection upon success and only then issue the request. It will then be sent on the only remaining, direct connection

That seems the best way to do it, thanks ! :)

**Reply by thomaseizinger** [reply]
Testing hole punching is a bit tricky. I am working on it at the moment. See https://github.com/libp2p/test-plans/issues/116 and https://github.com/libp2p/test-plans/issues/126.

**Reply by mkuhn19** [reply]
Ok I will follow it with great interest. Thanks again !

**Comment by dariusc93** [comment]
>     3. I am using request response to send message from a peer to another. When calling send_request, it tries to dial the remote peer and when there is a connection, it sends a message, as described [here](https://docs.rs/libp2p-request-response/latest/libp2p_request_response/struct.Behaviour.html#method.send_request). The issue is that, when a relayed circuit is established between 2 peers, the request will be sent directly and not wait until hole punching is performed. Can I tell request response to wait before sending the message ?

If youre polling swarm itself, you can wait for the event there thus queuing your message until there is a connection and a hole punch (if applicable), however for me i would rather just send the message through the relay regardless of if the hole punch or not since if it does, it will eventually upgrade the connections but if not, it would still utilize the relay, assuming the data is below the limit on the relay itself.

**Reply by mkuhn19** [reply]
Would you still do it if the peer only needs to send one request and then close the connection to the remote ?

**Reply by dariusc93** [reply]
If I am awaiting for an acknowledgement of the request or a response the connection would remain open usually and if i dont need anything else, i would close the connection if need be.
